File: src/observability/sink.py
"""Observability sinks (Python twin of sink.ts).

The ``JsonlFileSink`` body is the crash-safe O_APPEND / single-write-per-line /
daily-local-date pattern adapted directly from ``src/event_log.py``. A sink MUST
be best-effort (never raise out). ``otlp-http`` is NOT built here -- only the
``Sink`` protocol it will later implement.
"""

from __future__ import annotations

import json
import sys
import time
from pathlib import Path
from typing import Any, Protocol

# `workspace_default` is a flat module at src/; importable because the same src/
# on sys.path that makes the `observability` package importable also resolves it.
from workspace_default import resolve_workspace

__all__ = ["Sink", "JsonlFileSink", "sink_from_config"]


class Sink(Protocol):
    type: str

    def write(self, ev: dict[str, Any]) -> None: ...


class JsonlFileSink:
    type = "jsonl-file"

    def __init__(self, dir: Path | None = None) -> None:
        self._dir = dir

    def write(self, ev: dict[str, Any]) -> None:
        try:
            d = self._dir if self._dir is not None else (resolve_workspace(migrate=False) / "logs")
            d.mkdir(parents=True, exist_ok=True)
            ts = float(ev.get("ts", time.time()))
            date = time.strftime("%Y-%m-%d", time.localtime(ts))
            path = d / f"events-{date}.jsonl"
            line = json.dumps(ev, ensure_ascii=False, separators=(",", ":")) + "\n"
            # Single write() for atomicity on POSIX (event_log.py contract).
            with path.open("ab") as fh:
                fh.write(line.encode("utf-8"))
        except Exception as e:  # noqa: BLE001 — never raise out of a sink
            try:
                print(f"[obs] jsonl-file sink failed: {e}", file=sys.stderr)
            except Exception:
                pass


def sink_from_config(cfg: dict[str, Any]) -> Sink | None:
    t = cfg.get("type")
    if t == "jsonl-file":
        p = cfg.get("path")
        return JsonlFileSink(Path(p) if isinstance(p, str) else None)
    try:
        print(f"[obs] sink type {t!r} not supported yet, skipping", file=sys.stderr)
    except Exception:
        pass
    return None

File: src/observability/sink.ts
/**
 * Observability sinks. A `Sink` accepts a finished `ObsEvent` and ships it
 * somewhere; it MUST be best-effort (never throw out — `emit()` isolates each
 * sink, but a sink that swallows its own errors keeps the others clean).
 *
 * The `Sink` interface stays in `src/observability` permanently. The concrete
 * `JsonlFileSink` body is adapted from `src/event_log.py` (crash-safe O_APPEND,
 * single write per line, daily local-date file). In a later phase the concrete
 * sinks may move to their own file; the interface does not move. The
 * `otlp-http` sink is NOT built here — only the interface it will implement.
 *
 * Twin of sink.py.
 */

import { appendFileSync, mkdirSync } from 'node:fs';
import { join } from 'node:path';
import { resolveWorkspace } from '../workspace_default.js';
import type { ObsEvent } from './events.js';
import type { SinkConfig } from './config.js';

export interface Sink {
	readonly type: string;
	write(ev: ObsEvent): void; // best-effort; must not throw out
}

function dailyFilePath(dir: string, atMs: number): string {
	const d = new Date(atMs);
	const y = d.getFullYear();
	const m = String(d.getMonth() + 1).padStart(2, '0');
	const day = String(d.getDate()).padStart(2, '0');
	return join(dir, `events-${y}-${m}-${day}.jsonl`);
}

/** Append-only JSONL sink: `<dir>/events-YYYY-MM-DD.jsonl` (default dir
 *  `<workspace>/logs`). One compact JSON object per line, atomic O_APPEND. */
export class JsonlFileSink implements Sink {
	readonly type = 'jsonl-file';
	private readonly dir: string;

	constructor(opts?: { dir?: string }) {
		this.dir = opts?.dir ?? join(resolveWorkspace(), 'logs');
	}

	write(ev: ObsEvent): void {
		try {
			mkdirSync(this.dir, { recursive: true });
			const line = JSON.stringify(ev) + '\n';
			const path = dailyFilePath(this.dir, ev.ts * 1000);
			appendFileSync(path, line, { flag: 'a' });
		} catch (e) {
			try {
				process.stderr.write(`[obs] jsonl-file sink failed to write ${ev.kind}: ${(e as Error).message}\n`);
			} catch {
				/* even the warn is best-effort */
			}
		}
	}
}

/** Build a concrete sink from a config entry. Returns null for an unsupported
 *  type (e.g. otlp-http, not built yet) with a one-line warn. */
export function sinkFromConfig(cfg: SinkConfig): Sink | null {
	switch (cfg.type) {
		case 'jsonl-file':
			return new JsonlFileSink({ dir: typeof cfg.path === 'string' ? cfg.path : undefined });
		default:
			try {
				process.stderr.write(`[obs] sink type '${cfg.type}' not supported yet, skipping\n`);
			} catch {
				/* best-effort */
			}
			return null;
	}
}
